central/networkgraph/flow/datastore/internal/store/postgres/tests/store_test.go

```diff
@@ -254,24 +255,105 @@ func (s *NetworkflowStoreSuite) TestPruneStaleNetworkFlows() {
 		},
 	}
 
-	err := s.store.UpsertFlows(s.ctx, flows, timestamp.Now())
+	err := s.flowStore.UpsertFlows(s.ctx, flows, timestamp.Now())
 	s.Nil(err)
 
-	row := s.pool.QueryRow(s.ctx, flowsCountStmt)
+	row := s.pgDB.DB.QueryRow(s.ctx, flowsCountStmt)
 	var count int
 	err = row.Scan(&count)
 	s.Nil(err)
 	s.Equal(count, len(flows))
 
-	err = s.store.RemoveStaleFlows(s.ctx)
+	err = s.flowStore.RemoveStaleFlows(s.ctx)
 	s.Nil(err)
 
-	row = s.pool.QueryRow(s.ctx, flowsCountStmt)
+	row = s.pgDB.DB.QueryRow(s.ctx, flowsCountStmt)
 	err = row.Scan(&count)
 	s.Nil(err)
 	s.Equal(count, 2)
 }
 
+func (s *NetworkflowStoreSuite) TestPruneOrphanedExternalEntities() {
+	s.T().Setenv(features.ExternalIPs.EnvVar(), "true")
+
+	now := time.Now()
+
+	extEntity1 := ngTestutils.GetDiscoveredExtSrcNetworkEntity("223.42.0.1/32", clusterID)
+	err := s.entityStore.UpdateExternalNetworkEntity(s.ctx, extEntity1, false)
+	s.Nil(err)
+
+	extEntity2 := ngTestutils.GetDiscoveredExtSrcNetworkEntity("223.42.0.2/32", clusterID)
+	err = s.entityStore.UpdateExternalNetworkEntity(s.ctx, extEntity2, false)
+	s.Nil(err)
+
+	flows := []*storage.NetworkFlow{
+		{
+			Props: &storage.NetworkFlowProperties{
+				DstPort: 22,
+				DstEntity: &storage.NetworkEntityInfo{
+					Type: storage.NetworkEntityInfo_DEPLOYMENT,
+					Id:   fixtureconsts.Deployment1,
+				},
+				SrcEntity: &storage.NetworkEntityInfo{
+					Type: storage.NetworkEntityInfo_EXTERNAL_SOURCE,
+					Id:   extEntity1.GetInfo().Id,
+				},
+			},
+			ClusterId:         clusterID,
+			LastSeenTimestamp: timestamppb.New(now.Add(-100 * time.Second)),
+		},
+		{
+			Props: &storage.NetworkFlowProperties{
+				DstPort: 22,
+				DstEntity: &storage.NetworkEntityInfo{
+					Type: storage.NetworkEntityInfo_EXTERNAL_SOURCE,
+					Id:   extEntity2.GetInfo().Id,
+				},
+				SrcEntity: &storage.NetworkEntityInfo{
+					Type: storage.NetworkEntityInfo_DEPLOYMENT,
+					Id:   fixtureconsts.Deployment1,
+				},
+			},
+			ClusterId:         clusterID,
+			LastSeenTimestamp: timestamppb.New(now.Add(-100 * time.Second)),
+		},
+	}
+
+	err = s.flowStore.UpsertFlows(s.ctx, flows, timestamp.FromGoTime(now.Add(-100*time.Second)))
+	s.Nil(err)
+
+	// flows initially in the DB
+	row := s.pgDB.DB.QueryRow(s.ctx, flowsCountStmt)
+	var count int
+	err = row.Scan(&count)
+	s.Nil(err)
+	s.Equal(len(flows), count)
+
+	// entities initially in the DB
+	row = s.pgDB.DB.QueryRow(s.ctx, entitiesCountStmt)
+	err = row.Scan(&count)
+	s.Nil(err)
+	s.Equal(2, count)
+
+	// pruning (anything older than 10s).
+	// Flows should get pruned because Deployment1 is not in the DB.
+	window := now.Add(-10 * time.Second)
+	err = s.flowStore.RemoveOrphanedFlows(s.ctx, &window)
+	s.Nil(err)
+
+	// flows after pruning
+	row = s.pgDB.DB.QueryRow(s.ctx, flowsCountStmt)
+	err = row.Scan(&count)
+	s.Nil(err)
+	s.Equal(0, count)
+
+	// entities after pruning
+	row = s.pgDB.DB.QueryRow(s.ctx, entitiesCountStmt)
+	err = row.Scan(&count)
+	s.Nil(err)
+	s.Equal(0, count)
+}
```

Comment:
More test cases are needed. There should be a test where there are two flows that have the same entity and one of the flows is pruned and the other is not. The test should make sure that the entity is not pruned. There should also be a test where no flows are pruned. These test cases can be put into a loop.

Reply:
I added 2 more tests for the flows pruning (https://github.com/stackrox/stackrox/pull/14769/commits/20debf8a56cb97e80a4cdbacc73f7ae7cde7af3d), and 1 more for the case where we delete a deployment (https://github.com/stackrox/stackrox/pull/14769/commits/cb5f54d8ab0cae2b016700f2a6a9b7083cf25b5d).